File: Sources/KakaoSapiens/Services/SyncEnrollmentBuilder.swift
import CryptoKit
import Foundation

struct SyncEnrollmentPackage {
    let accountID: String
    let deviceID: String
    let enrollmentID: String
    let recoveryPhrase: String
    let secrets: SyncSecretBundle
    let rawRequestBody: Data
}

enum SyncEnrollmentBuilder {
    static func build(
        accountID: String,
        deviceID: String,
        enrollmentID: String,
        spaceID: String,
        platform: String,
        accountMasterKey: Data,
        deviceToken: Data,
        recoveryEntropy: Data,
        recoveryNonce: Data,
        words: [String]
    ) throws -> SyncEnrollmentPackage {
        let secrets = try SyncSecretBundle(accountMasterKey: accountMasterKey, deviceToken: deviceToken)
        let recovery = try SyncE2EE.deriveRecoveryMaterial(recoveryEntropy: recoveryEntropy)
        let verifier = try SyncE2EE.recoveryAuthVerifier(recovery.recoveryAuth)
        let context = SyncE2EE.RecoveryAADContext(
            accountID: accountID,
            recoveryLookup: recovery.recoveryLookup,
            recoveryVersion: 1
        )
        let wrapped = try SyncE2EE.sealRecoveryWrappedMasterKey(
            accountMasterKey: accountMasterKey,
            recoveryWrapKey: recovery.recoveryWrapKey,
            nonce: recoveryNonce,
            context: context
        )
        let tokenHash = Data(SHA256.hash(data: deviceToken)).map {
            String(format: "%02x", $0)
        }.joined()
        let json: [String: Any] = [
            "protocol_version": 1,
            "enrollment_id": enrollmentID,
            "account_id": accountID,
            "device": [
                "device_id": deviceID,
                "space_id": spaceID,
                "platform": platform,
                "display_name": NSNull(),
                "device_token_hash": tokenHash,
            ],
            "recovery": [
                "recovery_version": 1,
                "recovery_lookup": recovery.recoveryLookup.base64EncodedString(),
                "recovery_auth_verifier": verifier.base64EncodedString(),
                "wrapped_master_key": wrapped.base64EncodedString(),
            ],
        ]
        let raw = try JSONSerialization.data(withJSONObject: json, options: [.sortedKeys])
        return SyncEnrollmentPackage(
            accountID: accountID,
            deviceID: deviceID,
            enrollmentID: enrollmentID,
            recoveryPhrase: try SyncRecoveryMnemonic.encode(entropy: recoveryEntropy, words: words),
            secrets: secrets,
            rawRequestBody: raw
        )
    }
}

File: android/app/src/main/java/com/sapiens/gagaodok/sync/SyncEnrollmentBuilder.kt
package com.sapiens.gagaodok.sync

import java.security.MessageDigest
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

data class SyncEnrollmentPackage(
    val accountId:String,val deviceId:String,val enrollmentId:String,val recoveryPhrase:String,
    val secrets:SyncSecretBundle,val rawRequestBody:ByteArray,
)
object SyncEnrollmentBuilder {
 fun build(accountId:String,deviceId:String,enrollmentId:String,spaceId:String,platform:String,accountMasterKey:ByteArray,deviceToken:ByteArray,recoveryEntropy:ByteArray,recoveryNonce:ByteArray,words:List<String>):SyncEnrollmentPackage{
  val secrets=SyncSecretBundle(accountMasterKey,deviceToken);val recovery=SyncE2EE.deriveRecoveryMaterial(recoveryEntropy);val verifier=SyncE2EE.recoveryAuthVerifier(recovery.recoveryAuth);val wrapped=SyncE2EE.sealRecoveryWrappedMasterKey(accountMasterKey,recovery.recoveryWrapKey,recoveryNonce,SyncE2EE.RecoveryAADContext(accountId,recovery.recoveryLookup,1));val tokenHash=MessageDigest.getInstance("SHA-256").digest(deviceToken).joinToString(""){"%02x".format(it)}
  val body=buildJsonObject{put("protocol_version",1);put("enrollment_id",enrollmentId);put("account_id",accountId);put("device",buildJsonObject{put("device_id",deviceId);put("space_id",spaceId);put("platform",platform);put("display_name",JsonNull);put("device_token_hash",tokenHash)});put("recovery",buildJsonObject{put("recovery_version",1);put("recovery_lookup",SyncE2EE.encodeBase64(recovery.recoveryLookup));put("recovery_auth_verifier",SyncE2EE.encodeBase64(verifier));put("wrapped_master_key",SyncE2EE.encodeBase64(wrapped))})}
  return SyncEnrollmentPackage(accountId,deviceId,enrollmentId,SyncRecoveryMnemonic.encode(recoveryEntropy,words),secrets,body.toString().toByteArray())
 }
}
